File: python/src/ui/interactive_menu.py
"""Interactive CLI menu for the Choptyuk verification suite.

Provides a full interactive menu with:
- Run full verification
- Run simulations
- Configure all parameters
- Build custom hypotheses
- Generate reports in any format
- Generate plots
- View results
- Save/load configurations
"""

from __future__ import annotations
import json
import sys
import os
from pathlib import Path
from typing import Dict, Optional
import numpy as np
import logging

from ..core.klein_curve import KleinCurve
from ..core.spinor_phases import SpinorPhases
from ..core.dirac_operator import DiracOperator
from ..core.choptyuk_formula import ChoptyukFormula
from ..core.surfaces import SurfaceSpec, BOLZA, BRING, MACBEATH
from ..core.qnm import QNMPredictor
from ..core.hypothesis import HypothesisTester, HypothesisConfig
from ..verification.verify_all import VerificationSuite
from ..simulation.simulator import Simulator
from ..visualization.plots import PlotGenerator
from ..reporting.report_writer import ReportWriter

logger = logging.getLogger(__name__)


class InteractiveMenu:
    """Full interactive CLI menu system."""

    def __init__(self, output_base: str = "output"):
        self.output_base = Path(output_base)
        self.output_base.mkdir(parents=True, exist_ok=True)

        # Default parameters (all customizable)
        self.params = {
            "genus": 3, "K": -1.0, "psl_order": 168, "lambda_1": 3.838,
            "delta_A": None, "delta_B": None, "delta_C": None,
            "k_struct": 22, "c4": 0.125, "c6": 0.5,
            "delta_obs": 3.443, "b_ch_obs": 0.377,
            "dpi": 600, "precision": 15,
        }
        self.results = None
        self.sim_results = None
        self.logs = ""
        self.plot_paths = []
        self.report_paths = {}

    def run(self):
        """Main menu loop."""
        while True:
            self._print_header()
            choice = input("\n  Select option [1-10]: ").strip()
            if choice == "1":
                self._run_verification()
            elif choice == "2":
                self._run_simulation()
            elif choice == "3":
                self._configure_params()
            elif choice == "4":
                self._custom_hypothesis()
            elif choice == "5":
                self._generate_reports()
            elif choice == "6":
                self._generate_plots()
            elif choice == "7":
                self._view_results()
            elif choice == "8":
                self._load_preset()
            elif choice == "9":
                self._save_config()
            elif choice == "10" or choice.lower() == "q":
                print("\n  Exiting. All outputs saved to:", self.output_base)
                break
            else:
                print("  Invalid choice. Try again.")

    def _print_header(self):
        print("\n" + "=" * 60)
        print("  CHOPTYUK SPINOR CORRECTIONS - Interactive Menu")
        print("=" * 60)
        print("  1. Run Full Verification")
        print("  2. Run Simulations (parameter sweeps)")
        print("  3. Configure Parameters")
        print("  4. Build Custom Hypothesis")
        print("  5. Generate Reports (docx/pdf/txt/md/csv/html/json)")
        print("  6. Generate Plots (600 DPI PNG + PDF + SVG)")
        print("  7. View Results")
        print("  8. Load Preset Configuration")
        print("  9. Save Current Configuration")
        print("  10. Exit")

    def _run_verification(self):
        print("\n--- Running Full Verification ---")
        suite = VerificationSuite(
            genus=self.params["genus"], K=self.params["K"],
            psl_order=self.params["psl_order"], lambda_1=self.params["lambda_1"],
            delta_A=self.params["delta_A"], delta_B=self.params["delta_B"],
            delta_C=self.params["delta_C"], k_struct=self.params["k_struct"],
            c4=self.params["c4"], c6=self.params["c6"],
            delta_obs=self.params["delta_obs"], b_ch_obs=self.params["b_ch_obs"],
        )
        self.results = suite.run()
        self.logs = suite.get_logs()
        print("  Verification complete.")
        ch = self.results.get("choptyuk", {})
        print(f"  Delta_bC = {ch.get('delta_bc', 0):.6f} (dev {ch.get('deviation_bc_pct', 0):.3f}%)")
        print(f"  Delta_Ch (full) = {ch.get('delta_ch_full', 0):.6f} (dev {ch.get('deviation_full_pct', 0):.3f}%)")

    def _run_simulation(self):
        print("\n--- Running Simulations ---")
        sim = Simulator(delta_obs=self.params["delta_obs"])

        print("  1. Sweep delta_C")
        print("  2. Sweep lambda_1")
        print("  3. Convergence analysis")
        print("  4. Run all simulations")
        choice = input("  Select [1-4]: ").strip()

        dC = self.params["delta_C"] or np.pi / 7
        if choice in ("1", "4"):
            print("  Sweeping delta_C [0.1, 1.0]...")
            sim.sweep_delta_C(n_points=200, lambda_D2=self.params["lambda_1"] + self.params["K"])
        if choice in ("2", "4"):
            print("  Sweeping lambda_1 [2.0, 6.0]...")
            sim.sweep_lambda_1(delta_C=dC, R=2 * self.params["K"])
        if choice in ("3", "4"):
            print("  Convergence analysis (order 1..10)...")
            sim.convergence_analysis(max_order=10, delta_C=dC)

        self.sim_results = sim.results
        self.logs += "\n" + sim.get_logs()
        print("  Simulations complete.")

    def _configure_params(self):
        print("\n--- Configure Parameters ---")
        print("  Current values:")
        for k, v in self.params.items():
            print(f"    {k} = {v}")
        print("\n  Enter parameter name and value (e.g., 'lambda_1 4.0')")
        print("  Enter 'done' to finish, 'reset' for defaults")
        while True:
            inp = input("  > ").strip()
            if inp.lower() == "done":
                break
            if inp.lower() == "reset":
                self.params = {
                    "genus": 3, "K": -1.0, "psl_order": 168, "lambda_1": 3.838,
                    "delta_A": None, "delta_B": None, "delta_C": None,
                    "k_struct": 22, "c4": 0.125, "c6": 0.5,
                    "delta_obs": 3.443, "b_ch_obs": 0.377,
                    "dpi": 600, "precision": 15,
                }
                print("  Reset to defaults.")
                break
            try:
                parts = inp.split()
                key = parts[0]
                if key in self.params:
                    val = float(parts[1]) if parts[1] not in ("None", "none") else None
                    if key in ("genus", "psl_order", "k_struct", "dpi", "precision"):
                        val = int(parts[1])
                    self.params[key] = val
                    print(f"    {key} = {val}")
                else:
                    print(f"    Unknown parameter: {key}")
            except (IndexError, ValueError):
                print("    Invalid input. Use: parameter_name value")

    def _custom_hypothesis(self):
        print("\n--- Build Custom Hypothesis ---")
        name = input("  Hypothesis name: ").strip() or "custom"
        desc = input("  Description: ").strip() or ""

        print("  Custom delta_C (enter value or 'default' for pi/7):")
        dC_input = input("  > ").strip()
        dC = None if dC_input in ("default", "") else float(dC_input)

        print("  Custom lambda_D2 (enter value or 'default' for 3.338):")
        lam_input = input("  > ").strip()
        lam = None if lam_input in ("default", "") else float(lam_input)

        print("  Custom k_struct (enter value or 'default' for 22):")
        k_input = input("  > ").strip()
        k = None if k_input in ("default", "") else int(k_input)

        config = HypothesisConfig(name=name, description=desc,
                                  custom_delta_C=dC,
                                  custom_lambda_D2=lam,
                                  custom_k_struct=k)
        tester = HypothesisTester(delta_obs=self.params["delta_obs"])
        result = tester.test_hypothesis(config)

        print(f"\n  Result: Delta_Ch = {result.delta_ch_full:.6f}")
        print(f"  Deviation = {result.deviation:.3f}%")
        print(f"  Status: {'PASS' if result.passed else 'FAIL'} (tolerance {tester.tolerance}%)")

        print("\n  Parameter sweep? (y/n):")
        if input("  > ").strip().lower() == "y":
            print("  Sweep parameter (delta_C / lambda_D2 / k_struct):")
            param = input("  > ").strip()
            print("  Start value:")
            start = float(input("  > "))
            print("  End value:")
            end = float(input("  > "))
            print("  Number of points:")
            n = int(input("  > "))
            values = np.linspace(start, end, n).tolist()
            results = tester.parameter_sweep(param, values, config)
            print(f"  Sweep complete: {len(results)} points tested")
            best = min(results, key=lambda r: r.deviation)
            print(f"  Best: {best.name}, deviation = {best.deviation:.3f}%")

    def _generate_reports(self):
        if not self.results:
            print("  Run verification first (option 1).")
            return
        print("\n--- Generate Reports ---")
        print("  Formats: docx, pdf, txt, md, csv, html, json")
        fmt_input = input("  Enter formats (comma-separated, or 'all'): ").strip()
        if fmt_input.lower() in ("all", ""):
            formats = ["docx", "pdf", "txt", "md", "csv", "html", "json"]
        else:
            formats = [f.strip() for f in fmt_input.split(",")]

        writer = ReportWriter(str(self.output_base / "reports"), formats)
        self.report_paths = writer.generate_all(self.results, self.logs)
        print(f"  Generated {len(self.report_paths)} reports:")
        for fmt, path in self.report_paths.items():
            print(f"    {fmt}: {path}")

    def _generate_plots(self):
        print("\n--- Generate Plots ---")
        plotter = PlotGenerator(str(self.output_base / "plots"), self.params["dpi"])
        self.plot_paths = plotter.generate_all(
            self.results or {}, self.sim_results
        )
        print(f"  Generated {len(self.plot_paths)} plot files:")
        for p in self.plot_paths:
            print(f"    {p}")

    def _view_results(self):
        if not self.results:
            print("  No results yet. Run verification first.")
            return
        print("\n--- Results Summary ---")
        ch = self.results.get("choptyuk", {})
        print(f"  Delta_bC = {ch.get('delta_bc', 0):.6f} (dev {ch.get('deviation_bc_pct', 0):.3f}%)")
        print(f"  Delta_Ch (base) = {ch.get('delta_ch_base', 0):.6f} (dev {ch.get('deviation_ch_pct', 0):.3f}%)")
        print(f"  Delta_Ch (full) = {ch.get('delta_ch_full', 0):.6f} (dev {ch.get('deviation_full_pct', 0):.3f}%)")
        print(f"  b_Ch = {ch.get('b_ch', 0):.6f} (dev {ch.get('deviation_b_ch_pct', 0):.3f}%)")
        if self.plot_paths:
            print(f"\n  Plot files: {len(self.plot_paths)}")
        if self.report_paths:
            print(f"  Report files: {len(self.report_paths)}")

    def _load_preset(self):
        print("\n--- Load Preset ---")
        print("  Available: standard, high_precision, ligo_analysis")
        name = input("  Preset name: ").strip()
        preset_dir = Path(__file__).parent.parent.parent / "presets"
        preset_file = preset_dir / f"{name}.json"
        if preset_file.exists():
            with open(preset_file) as f:
                preset = json.load(f)
            print(f"  Loaded preset: {preset.get('name', name)}")
            print(f"  Description: {preset.get('description', '')}")
        else:
            print(f"  Preset not found: {preset_file}")

    def _save_config(self):
        print("\n--- Save Configuration ---")
        name = input("  Config name: ").strip() or "custom"
        config_path = self.output_base / f"config_{name}.json"
        with open(config_path, 'w') as f:
            json.dump(self.params, f, indent=2, default=str)
        print(f"  Saved to: {config_path}")

File: julia/src/interactive_menu.jl
"""
    Interactive REPL menu for the Choptyuk spinor monograph.

Provides a full interactive menu with options for running verification,
simulation, configuring parameters, generating reports and plots,
and more.
"""
module InteractiveMenu

using ..ChoptyukSpinor
using Printf
using JSON

"""
    MenuState

State of the interactive menu session.

# Fields
- `simulator::Simulator`: Active simulator instance
- `running::Bool`: Whether the menu loop is active
- `output_dir::String`: Output directory for reports and plots
"""
mutable struct MenuState
    simulator::Simulator
    running::Bool
    output_dir::String
end

"""
    MenuState(; kwargs...) -> MenuState

Construct a menu state with default or custom parameters.
"""
function MenuState(; output_dir::String = "output", kwargs...)
    sim = Simulator(; kwargs...)
    return MenuState(sim, true, output_dir)
end

"""
    print_header()

Print the menu header.
"""
function print_header()
    println()
    println("=" ^ 70)
    println("  CHOPTYUK SPINOR MONOGRAPH — Interactive Verification System")
    println("  Spinor corrections b-C and a-C on the Klein quartic curve")
    println("=" ^ 70)
    println()
end

"""
    print_menu()

Print the main menu options.
"""
function print_menu()
    println("Main Menu:")
    println("-" ^ 40)
    println("  1. Run verification")
    println("  2. Run simulation")
    println("  3. Configure parameters")
    println("  4. Custom hypothesis")
    println("  5. Generate reports")
    println("  6. Generate plots")
    println("  7. View results")
    println("  8. Load preset")
    println("  9. Save config")
    println("  0. Exit")
    println("-" ^ 40)
end

"""
    get_input(prompt::String; default::String="") -> String

Get user input with an optional default value.
"""
function get_input(prompt::String; default::String = "")
    if isempty(default)
        print("$prompt: ")
    else
        print("$prompt [$default]: ")
    end
    input = readline()
    input = strip(input)
    return isempty(input) ? default : input
end

"""
    get_float(prompt::String; default::Float64=0.0) -> Float64

Get a float from user input.
"""
function get_float(prompt::String; default::Float64 = 0.0)
    input = get_input(prompt; default = string(default))
    try
        return parse(Float64, input)
    catch
        println("  Invalid number, using default: $default")
        return default
    end
end

"""
    get_int(prompt::String; default::Int=0) -> Int

Get an integer from user input.
"""
function get_int(prompt::String; default::Int = 0)
    input = get_input(prompt; default = string(default))
    try
        return parse(Int, input)
    catch
        println("  Invalid number, using default: $default")
        return default
    end
end

"""
    run_verification(state::MenuState)

Run all verification checks and display results.
"""
function run_verification(state::MenuState)
    println("\n--- Running Verification ---")
    cfg = state.simulator.config

    # Klein curve
    println("\n[1] Klein Curve:")
    K = KleinCurve(lambda_1 = cfg.lambda_1, R = Int(cfg.R))
    kv = verify_relations(K)
    all_pass = true
    for (key, val) in sort(collect(pairs(kv)))
        status = val ? "PASS" : "FAIL"
        val || (all_pass = false)
        @printf("  [%s] %s\n", status, key)
    end
    @printf("\n  Overall: %s\n", all_pass ? "ALL PASS" : "SOME FAILURES")

    # Spinor phases
    println("\n[2] Spinor Phases:")
    phases = SpinorPhases(delta_C = cfg.delta_C)
    pv = verify_phase_relations(phases)
    all_pass = true
    for (key, val) in sort(collect(pairs(pv)))
        status = val ? "PASS" : "FAIL"
        val || (all_pass = false)
        @printf("  [%s] %s\n", status, key)
    end
    @printf("\n  Overall: %s\n", all_pass ? "ALL PASS" : "SOME FAILURES")

    # Dirac operator
    println("\n[3] Dirac Operator:")
    D = DiracOperator(lambda_D2_triv = cfg.lambda_D2_triv, lambda_1 = cfg.lambda_1, R = cfg.R)
    dv = verify_dirac_relations(D)
    all_pass = true
    for (key, val) in sort(collect(pairs(dv)))
        status = val ? "PASS" : "FAIL"
        val || (all_pass = false)
        @printf("  [%s] %s\n", status, key)
    end
    @printf("\n  Overall: %s\n", all_pass ? "ALL PASS" : "SOME FAILURES")

    # Choptyuk formula
    println("\n[4] Choptyuk Formula:")
    cf = ChoptyukFormula(lambda_D2_triv = cfg.lambda_D2_triv, delta_C = cfg.delta_C)
    cv = verify_choptyuk_formula(cf; observed = cfg.observed_delta)
    for (key, val) in sort(collect(pairs(cv)))
        if isa(val, Bool)
            @printf("  [%s] %s\n", val ? "PASS" : "FAIL", key)
        else
            @printf("  %s = %.6f\n", key, Float64(val))
        end
    end

    # QNM
    println("\n[5] QNM Predictions:")
    qv = verify_qnm()
    for (key, val) in sort(collect(pairs(qv)))
        @printf("  [%s] %s\n", val ? "PASS" : "FAIL", key)
    end

    # Surfaces
    println("\n[6] Surface Verification:")
    for S in ALL_SURFACES
        sv = verify_surface(S)
        n_pass = count(values(sv))
        @printf("  %s: %d/%d checks passed\n", S.name, n_pass, length(sv))
    end

    println("\n--- Verification Complete ---")
end

"""
    run_simulation(state::MenuState)

Run the full simulation.
"""
function run_simulation(state::MenuState)
    println("\n--- Running Full Simulation ---")
    results = run_full_simulation(state.simulator)
    println("  Simulation complete.")
    @printf("  Results keys: %s\n", join(sort(collect(keys(results))), ", "))
    @printf("  Log entries: %d\n", length(state.simulator.log))
    println("\n--- Simulation Complete ---")
end

"""
    configure_parameters(state::MenuState)

Interactively configure simulation parameters.
"""
function configure_parameters(state::MenuState)
    cfg = state.simulator.config

    println("\n--- Configure Parameters ---")
    println("  Current values shown in brackets. Press Enter to keep.")

    cfg.lambda_D2_triv = get_float("  λ_{D²,triv}"; default = cfg.lambda_D2_triv)
    cfg.lambda_1 = get_float("  λ₁ (Laplacian eigenvalue)"; default = cfg.lambda_1)
    cfg.delta_C = get_float("  δ_C (spinor phase)"; default = cfg.delta_C)
    cfg.R = get_float("  R (scalar curvature)"; default = cfg.R)
    cfg.observed_delta = get_float("  Δ_obs (observed)"; default = cfg.observed_delta)
    cfg.tolerance = get_float("  Tolerance"; default = cfg.tolerance)
    cfg.n_sweep = get_int("  N_sweep (sweep points)"; default = cfg.n_sweep)

    println("\n  Parameters updated.")
    println("--- Configuration Complete ---")
end

"""
    custom_hypothesis(state::MenuState)

Run a custom hypothesis test with user-specified parameters.
"""
function custom_hypothesis(state::MenuState)
    println("\n--- Custom Hypothesis Test ---")

    lambda_D2 = get_float("  λ_{D²,triv}"; default = state.simulator.config.lambda_D2_triv)
    delta_C = get_float("  δ_C"; default = state.simulator.config.delta_C)
    observed = get_float("  Δ_obs"; default = state.simulator.config.observed_delta)
    tolerance = get_float("  Tolerance"; default = state.simulator.config.tolerance)

    tester = HypothesisTester(
        lambda_D2_triv = lambda_D2,
        delta_C = delta_C,
        observed_delta = observed,
        tolerance = tolerance,
    )
    result = test_hypothesis(tester)

    @printf("\n  Results:\n")
    @printf("    Δ_Ch = %.6f\n", result.delta_ch)
    @printf("    Δ_obs = %.6f\n", result.delta_obs)
    @printf("    Deviation = %.6f\n", result.deviation)
    @printf("    Relative deviation = %.6f\n", result.relative_deviation)
    status = result.accepted ? "ACCEPTED" : "REJECTED"
    @printf("    Hypothesis: %s (tolerance = %.6f)\n", status, result.tolerance)

    # Sensitivity analysis
    sens = sensitivity_analysis(tester)
    println("\n  Sensitivity Analysis:")
    for (param, val) in sort(collect(pairs(sens)))
        @printf("    ∂Δ_Ch/∂%s = %.6f\n", param, val)
    end

    println("\n--- Custom Hypothesis Complete ---")
end

"""
    generate_reports(state::MenuState)

Generate reports in all formats.
"""
function generate_reports(state::MenuState)
    println("\n--- Generating Reports ---")

    if isempty(state.simulator.results)
        println("  No results yet. Running simulation first...")
        run_full_simulation(state.simulator)
    end

    Reporting.generate_all_reports(
        state.simulator.results,
        state.simulator.log;
        dir = state.output_dir,
    )

    println("  Reports saved to: $(state.output_dir)")
    println("--- Reports Complete ---")
end

"""
    generate_plots(state::MenuState)

Generate all visualization plots.
"""
function generate_plots(state::MenuState)
    println("\n--- Generating Plots ---")

    if isempty(state.simulator.results)
        println("  No results yet. Running simulation first...")
        run_full_simulation(state.simulator)
    end

    Visualization.generate_all_plots(
        state.simulator.results;
        dir = state.output_dir,
    )

    println("  Plots saved to: $(state.output_dir)")
    println("--- Plots Complete ---")
end

"""
    view_results(state::MenuState)

View the current results.
"""
function view_results(state::MenuState)
    println("\n--- Current Results ---")

    if isempty(state.simulator.results)
        println("  No results yet. Run verification or simulation first.")
        return
    end

    for (key, val) in sort(collect(pairs(state.simulator.results)))
        if isa(val, HypothesisResult)
            @printf("  %s:\n", key)
            @printf("    Δ_Ch = %.6f, Δ_obs = %.6f\n", val.delta_ch, val.delta_obs)
            @printf("    Deviation = %.6f, %s\n", val.deviation, val.accepted ? "ACCEPTED" : "REJECTED")
        elseif isa(val, Dict{String, Bool})
            n_pass = count(values(val))
            @printf("  %s: %d/%d checks passed\n", key, n_pass, length(val))
        else
            @printf("  %s: (%s)\n", key, typeof(val))
        end
    end

    println("\n--- End of Results ---")
end

"""
    load_preset(state::MenuState)

Load a preset configuration.
"""
function load_preset(state::MenuState)
    println("\n--- Load Preset ---")
    println("  Available presets:")
    println("    1. standard (canonical Klein curve values)")
    println("    2. custom JSON file")

    choice = get_input("  Select preset"; default = "1")

    if choice == "1"
        # Look for standard.json in presets directory
        preset_path = joinpath(@__DIR__, "..", "presets", "standard.json")
        if isfile(preset_path)
            _load_preset_file(state, preset_path)
        else
            # Use built-in defaults
            println("  Loading built-in standard preset...")
            cfg = state.simulator.config
            cfg.lambda_D2_triv = 3.338
            cfg.lambda_1 = 3.838
            cfg.delta_C = π / 7
            cfg.R = -2.0
            cfg.observed_delta = 3.443
            cfg.tolerance = 0.01
            cfg.n_sweep = 100
            println("  Standard preset loaded.")
        end
    elseif choice == "2"
        filepath = get_input("  Enter preset file path")
        if isfile(filepath)
            _load_preset_file(state, filepath)
        else
            println("  File not found: $filepath")
        end
    else
        println("  Invalid choice.")
    end

    println("--- Preset Loading Complete ---")
end

"""
    _load_preset_file(state::MenuState, filepath::String)

Load a preset from a JSON file.
"""
function _load_preset_file(state::MenuState, filepath::String)
    try
        data = JSON.parsefile(filepath)
        cfg = state.simulator.config
        haskey(data, "lambda_D2_triv") && (cfg.lambda_D2_triv = Float64(data["lambda_D2_triv"]))
        haskey(data, "lambda_1") && (cfg.lambda_1 = Float64(data["lambda_1"]))
        haskey(data, "delta_C") && (cfg.delta_C = Float64(data["delta_C"]))
        haskey(data, "R") && (cfg.R = Float64(data["R"]))
        haskey(data, "observed_delta") && (cfg.observed_delta = Float64(data["observed_delta"]))
        haskey(data, "tolerance") && (cfg.tolerance = Float64(data["tolerance"]))
        haskey(data, "n_sweep") && (cfg.n_sweep = Int(data["n_sweep"]))
        println("  Preset loaded from: $filepath")
    catch e
        println("  Error loading preset: $e")
    end
end

"""
    save_config(state::MenuState)

Save the current configuration to a JSON file.
"""
function save_config(state::MenuState)
    println("\n--- Save Configuration ---")

    filepath = get_input("  Save to file"; default = joinpath(state.output_dir, "config.json"))

    cfg = state.simulator.config
    data = Dict(
        "lambda_D2_triv" => cfg.lambda_D2_triv,
        "lambda_1" => cfg.lambda_1,
        "delta_C" => cfg.delta_C,
        "R" => cfg.R,
        "observed_delta" => cfg.observed_delta,
        "tolerance" => cfg.tolerance,
        "n_sweep" => cfg.n_sweep,
    )

    try
        mkpath(dirname(filepath))
        write(filepath, JSON.json(data, 2))
        println("  Configuration saved to: $filepath")
    catch e
        println("  Error saving configuration: $e")
    end

    println("--- Save Complete ---")
end

"""
    run_interactive(; output_dir::String="output", kwargs...)

Run the interactive menu loop.
"""
function run_interactive(; output_dir::String = "output", kwargs...)
    state = MenuState(; output_dir = output_dir, kwargs...)

    print_header()

    while state.running
        print_menu()
        choice = get_input("Select option"; default = "0")

        try
            if choice == "1"
                run_verification(state)
            elseif choice == "2"
                run_simulation(state)
            elseif choice == "3"
                configure_parameters(state)
            elseif choice == "4"
                custom_hypothesis(state)
            elseif choice == "5"
                generate_reports(state)
            elseif choice == "6"
                generate_plots(state)
            elseif choice == "7"
                view_results(state)
            elseif choice == "8"
                load_preset(state)
            elseif choice == "9"
                save_config(state)
            elseif choice == "0" || lowercase(choice) == "exit"
                println("\nExiting. Goodbye!")
                state.running = false
            else
                println("  Invalid option: $choice")
            end
        catch e
            println("  Error: $e")
            @error "Menu error" exception = (e, catch_backtrace())
        end
    end
end

end # module InteractiveMenu
